crates/rust-analyzer/src/cli/lsif/lsif_types.rs

```diff
@@ -0,0 +1,354 @@
+            data: Element::Vertex(Vertex::ResultSet(ResultSet { key: None })),
+        };
+
+        let text = r#"{ "id": 2, "type": "vertex", "label": "resultSet" }"#.replace(' ', "");
+
+        assert_eq!(serde_json::to_string(&data).unwrap(), text);
+        assert_eq!(serde_json::from_str::<Entry>(&text).unwrap(), data);
+
+        let data = Entry {
+            id: lsp_types::NumberOrString::Number(4),
+            data: Element::Vertex(Vertex::ResultSet(ResultSet {
+                key: Some(String::from("hello")),
+            })),
+        };
+
+        let text = r#"{ "id": 4, "type": "vertex", "label": "resultSet", "key": "hello" }"#
+            .replace(' ', "");
+
+        assert_eq!(serde_json::to_string(&data).unwrap(), text);
+        assert_eq!(serde_json::from_str::<Entry>(&text).unwrap(), data);
+    }
+
+    #[test]
+    fn definition() {
+        let data = Entry {
+            id: lsp_types::NumberOrString::Number(21),
+            data: Element::Edge(Edge::Item(Item::Definition(EdgeData {
+                in_v: lsp_types::NumberOrString::Number(18),
+                out_v: lsp_types::NumberOrString::Number(16),
+            }))),
+        };
+
+        let text = r#"{ "id": 21, "type": "edge", "label": "item", "property": "definition", "outV": 16, "inV": 18}"#
+            .replace(' ', "");
+
+        assert_eq!(
+            serde_json::from_str::<serde_json::Value>(&text).unwrap(),
+            serde_json::to_value(&data).unwrap()
+        );
+    }
+
+    mod methods {
+        use super::*;
+
+        #[test]
+        fn references() {
+            let data = Entry {
+                id: lsp_types::NumberOrString::Number(17),
+                data: Element::Edge(Edge::References(EdgeData {
+                    in_v: lsp_types::NumberOrString::Number(16),
+                    out_v: lsp_types::NumberOrString::Number(15),
+                })),
+            };
+
+            let text = r#"{ "id": 17, "type": "edge", "label": "textDocument/references", "outV": 15, "inV": 16 }"#;
+
+            assert_eq!(
+                serde_json::from_str::<serde_json::Value>(&text).unwrap(),
+                serde_json::to_value(&data).unwrap()
+            );
+        }
+
+        #[test]
+        fn definition() {
+            let data = Entry {
+                id: lsp_types::NumberOrString::Number(13),
+                data: Element::Vertex(Vertex::DefinitionResult {
+                    result: DefinitionResultType::Scalar(LocationOrRangeId::RangeId(
+                        lsp_types::NumberOrString::Number(7),
+                    )),
+                }),
+            };
+
+            let text =
+                r#"{ "id": 13, "type": "vertex", "label": "definitionResult", "result": 7 }"#;
+
+            assert_eq!(
+                serde_json::from_str::<serde_json::Value>(&text).unwrap(),
+                serde_json::to_value(&data).unwrap()
+            );
+        }
+    }
+}
```

Comment:
These type definitions should live in a separate crate, to make sure that they are a separate compilation unit and can be compiled before we get to rust-analyzer crate. 

If these definitions are hand-written, they should live in another repository (probably makes sense to add them to `lsp-types` crates). 

If the definitions are generated from some sort of schema (need to check this, but I think lsif, unlike lsp, has schema somewhere), it's ok to keep them in the same repo, provided that the generating script is included as well.

Reply:
If `lsif_types` only contains type definitions, then I don't think it will need any changes if you implement all types in the LSIF spec at once. Looking at https://github.com/microsoft/lsif-node/blob/main/protocol/src/protocol.ts and ignoring all convenience methods this seems to be doable, though maybe a bit time consuming.

Reply:
Most of changes are for completing, but I one time had to change something (from the old code) because I was getting a runtime error for serialization. Although it won't be happen in a carefully tested implementation.

We can track a full implementation of `lsif_types` with tests in https://github.com/gluon-lang/lsp-types/issues/214 and in meantime, using the local one for not being blocked.

Reply:
I see how it's easier to iterate on the code if it just lives there, but, to keep the *rest* of the project more maintainable it's important that the code lives elsewhere, either in a separate crate, or in a separate repo. 

With a separate repo, the practical solution for fast iteration is the patch section of Cargo.toml: https://github.com/rust-analyzer/rust-analyzer/blob/1ef4b2cb7b84c45286b505512082273ccca41c57/Cargo.toml#L24-L33

Reply:
https://github.com/gluon-lang/lsp-types/pull/215 is now merged but it is not released in crates.io yet.